python/cuml/cuml/svm/svm_base.pyx

```diff
@@ -12,210 +12,148 @@
 
+cimport cython
 from libc.stdint cimport uintptr_t
 from libcpp cimport bool
 from pylibraft.common.handle cimport handle_t
 
-from cuml.internals.logger cimport level_enum
-from cuml.svm.kernel_params cimport KernelParams, KernelType
-
-
-cdef extern from "cuml/svm/svm_parameter.h" namespace "ML::SVM" nogil:
-    enum SvmType:
-        C_SVC,
-        NU_SVC,
-        EPSILON_SVR,
-        NU_SVR
-
-    cdef struct SvmParameter:
-        # parameters for training
-        double C
-        double cache_size
-        int max_iter
-        int nochange_steps
-        double tol
-        level_enum verbosity
-        double epsilon
-        SvmType svmType
-
-
-cdef extern from "cuml/svm/svm_model.h" namespace "ML::SVM" nogil:
-
-    cdef cppclass SupportStorage[math_t]:
-        int nnz
-        int* indptr
-        int* indices
-        math_t* data
-
-    cdef cppclass SvmModel[math_t]:
-        # parameters of a fitted model
-        int n_support
-        int n_cols
-        math_t b
-        math_t *dual_coefs
-        SupportStorage[math_t] support_matrix
-        int *support_idx
-        int n_classes
-        math_t *unique_labels
-
-cdef extern from "cuml/svm/svc.hpp" namespace "ML::SVM" nogil:
-
-    cdef void svcPredict[math_t](
-        const handle_t &handle, math_t* data, int n_rows, int n_cols,
-        KernelParams &kernel_params, const SvmModel[math_t] &model,
-        math_t *preds, math_t buffer_size, bool predict_class) except +
-
-    cdef void svcPredictSparse[math_t](
-        const handle_t &handle, int* indptr, int* indices,
-        math_t* data, int n_rows, int n_cols, int nnz,
-        KernelParams &kernel_params, const SvmModel[math_t] &model,
-        math_t *preds, math_t buffer_size, bool predict_class) except +
-
-    cdef void svmFreeBuffers[math_t](const handle_t &handle,
-                                     SvmModel[math_t] &m) except +
+cimport cuml.svm.svm_headers as lib
+
+
+@cython.no_gc_clear
+cdef class _SVMModel:
+    """A thin python wrapper for an SvmModel.
+
+    Used to keep the memory around until all references to
+    the underlying model are dropped."""
+    cdef lib.SvmModel[float] *model_f
+    cdef lib.SvmModel[double] *model_d
+    cdef object handle
+
+    @staticmethod
+    cdef _SVMModel new(handle, bool is_float32):
+        cdef _SVMModel self = _SVMModel.__new__(_SVMModel)
+        self.handle = handle
+        if is_float32:
+            self.model_f = new lib.SvmModel[float]()
+        else:
+            self.model_d = new lib.SvmModel[double]()
+        return self
+
+    def __dealloc__(self):
+        cdef handle_t* handle_ = <handle_t*><size_t>self.handle.getHandle()
+
+        if self.model_f != NULL:
+            lib.svmFreeBuffers(handle_[0], self.model_f[0])
+            del self.model_f
+            self.model_f = NULL
+        elif self.model_d != NULL:
+            lib.svmFreeBuffers(handle_[0], self.model_d[0])
```

Comment:
From looking at this function (and the overall `SvmModel` struct) I suspect the memory management on the C++ side here is outdated and could be refreshed a bit. I opted not to do that here and just work with the existing C++ api, but it might be good to have someone more versed on that side take a look for a followup.